### PR TITLE
Conform to IDNA version 15.1.0 revision 31

### PR DESCRIPTION
**Pull request by TRowbotham**
This adds the necessary changes to conform to [IDNA version 15.1.0 revision 31](https://www.unicode.org/reports/tr46/tr46-31.html#Modifications).

Notable Changes

* Transitional processing (the default in PHP) is now deprecated. No deprecation notices were added as PHP does not yet report a deprecation notice in this case.
* An error is no longer recorded for characters with a status of disallowed.
* When performing code point mapping and transitional processing is enabled the code point U+1E9E capital sharp s (ẞ), is replaced by the string “ss”
* A new internal option "IgnoreInvalidPunycode" was added, which is supposed to allow for an all-ASCII fast-path, however, the official tests do not test this configuration option.

### CODE REVIEW

**Comment by derrabus** [comment]
Can we add a test?

**Comment by stof** [comment]
I don't think we should add options that don't exist in PHP. This would defeat the purpose of this library.

**Comment by TRowbotham** [comment]
I have decided to remove the "IgnoreInvalidPunycode" option as suggested since the option is not available in PHP.

File: src/Intl/Idn/Idn.php [hunk 1]
```diff
@@ -346,6 +342,18 @@ private static function process($domain, array $options, Info $info)
...
+                // Step 4.1. If the label contains any non-ASCII code point (i.e., a code point greater than U+007F),
+                // record that there was an error, and continue with the next label.
+                if (1 === preg_match('/[^\x00-\x7F]/', $label)) {
+                    $info->errors |= self::ERROR_PUNYCODE;
+
+                    continue;
+                }
+
+                // Step 4.2. Attempt to convert the rest of the label to Unicode according to Punycode [RFC3492]. If
+                // that conversion fails and if not IgnoreInvalidPunycode, record that there was an error, and continue
```
Comment (stof) [inline src/Intl/Idn/Idn.php]: The reference to `IgnoreInvalidPunycode` should be removed

Comment (TRowbotham) [reply src/Intl/Idn/Idn.php]: Thanks for catching that

File: src/Intl/Idn/Idn.php [hunk 2]
```diff
@@ -294,7 +290,7 @@ private static function mapCodePoints($input, array $options, Info $info)
                     break;
 
                 case 'mapped':
-                    $str .= $data['mapping'];
+                    $str .= ($transitional && $codePoint === 0x1E9E ? 'ss' : $data['mapping']);
```
Comment (nicolas-grekas) [inline src/Intl/Idn/Idn.php]:
```suggestion
                    $str .= $transitional && 0x1E9E === $codePoint ? 'ss' : $data['mapping'];
```

File: src/Intl/Idn/Idn.php [hunk 3]
```diff
@@ -346,6 +342,18 @@ private static function process($domain, array $options, Info $info)
             $validationOptions = $options;
 
             if ('xn--' === substr($label, 0, 4)) {
+                // Step 4.1. If the label contains any non-ASCII code point (i.e., a code point greater than U+007F),
+                // record that there was an error, and continue with the next label.
+                if (1 === preg_match('/[^\x00-\x7F]/', $label)) {
```
Comment (nicolas-grekas) [inline src/Intl/Idn/Idn.php]:
```suggestion
                if (preg_match('/[^\x00-\x7F]/', $label)) {
```

File: src/Intl/Idn/Idn.php [hunk 4]
```diff
@@ -516,6 +524,8 @@ private static function validateLabel($label, Info $info, array $options, $canBe
             if ('-' === substr($label, -1, 1)) {
                 $info->errors |= self::ERROR_TRAILING_HYPHEN;
             }
+        } elseif (0 === strncmp($label, 'xn--', 4)) {
```
Comment (nicolas-grekas) [inline src/Intl/Idn/Idn.php]: `'xn--' === substr($label, 0, 4)` to match the same check above (or replace the code above maybe?)

**Review by nicolas-grekas — COMMENTED** [review]
Thanks for the PR.
Can you please add some tests?

File: src/Intl/Idn/Idn.php [hunk 5]
```diff
@@ -280,10 +280,6 @@ private static function mapCodePoints($input, array $options, Info $info)
 
             switch ($data['status']) {
                 case 'disallowed':
-                    $info->errors |= self::ERROR_DISALLOWED;
```
Comment (nicolas-grekas) [inline src/Intl/Idn/Idn.php:283]: Can you explain why this change?

**Review by nicolas-grekas — COMMENTED** [review]
I fixed my CS comments by force-pushing on your fork.
Tests would still be welcome. And I have one question (see inline comments).
Are you still interested in finishing this PR?

File: src/Intl/Idn/Idn.php (hunk 5 above)
Comment (TRowbotham) [reply src/Intl/Idn/Idn.php:283]: The reasoning for this change according to the spec is "Checking for disallowed characters both before and after normalization yielded inconsistent results for unnormalized vs. normalized text".

**Comment by TRowbotham** [comment]
> I fixed my CS comments by force-pushing on your fork. Tests would still be welcome. And I have one question (see inline comments). Are you still interested in finishing this PR?

Yes, sorry, haven't had a lot of free time lately.  I'll try to put some tests together this week.

Thanks for taking care of the style issues.

**Comment by nicolas-grekas** [comment]
Thank you @TRowbotham.

**Comment by nicolas-grekas** [comment]
@TRowbotham we have the CI fail because IDN tests don't work anymore. Could you please have a look?

**Comment by nicolas-grekas** [comment]
`./phpunit --filter Idn` BTW

**Comment by TRowbotham** [comment]
I'll try to take a look at this over the next couple of weeks. Since it doesn't look like anything has changed in the polyfill, my first guess would be that the the intl extension that PHP was built with in the test env is now using Unicode 16, which has a number of changes to it on the IDN side of things. Thanks for reaching out!

**Comment by xabbuh** [comment]
@TRowbotham I think it's quite likely that the intl extension version has changed as we also see other intl related failures (or have seen in the past) around date formats.

**Comment by TRowbotham** [comment]
I had some time to dig into this today, and the version of ICU being used by the intl extension seems to have changed in the test environment. This coupled with the test data being meant for Unicode version 13 means that the data and implementation are too far out of sync. Updating the unicode data and test data will require code changes to the IDN polyfill, which I am happy to make, but a quick fix would be pinning the ICU version for the intl extension. The last working ICU version was 73.2.

https://github.com/symfony/polyfill/blob/80b0d3fac4768e5987a2c39e34815efa60fd9662/.github/workflows/tests.yml#L37
```diff
- extensions: "apcu, intl, mbstring, odbc, uuid"
+ extensions: "apcu, intl-73.2, mbstring, odbc, uuid"
```

**Comment by nicolas-grekas** [comment]
Thanks for having a look. I'm fine with pinning, and upgrading separately.
